File: clients/python/baml_core/_impl/deserializer/enum_deserializer.py
import typing

from enum import Enum


from .base_deserialzier import (
    BaseDeserializer,
    CheckLutFn,
    RawWrapper,
    Result,
    Diagnostics,
    DeserializerError,
    DeserializerWarning,
)


T = typing.TypeVar("T", bound=Enum)


@typing.final
class EnumDeserializer(BaseDeserializer[T]):
    __enm: typing.Type[T]
    __value_aliases: typing.Dict[str, str]

    def __init__(
        self,
        *,
        enm: typing.Type[T],
        aliases: typing.Dict[str, str] = {},
    ):
        super().__init__(rank=5)
        self.__enm = enm
        # This field is alias to value.
        self.__value_aliases = aliases

    def copy_with_aliases(
        self, aliases: typing.Dict[str, typing.Optional[str]]
    ) -> "EnumDeserializer[T]":
        # Use any aliases that are defined in the model.
        _aliases = {
            key: value
            for key, value in self.__value_aliases.items()
            if aliases.get(key) is not None
        }
        # Update with any aliases that are defined in the call.
        _aliases.update(
            {key: value for key, value in aliases.items() if value is not None}
        )
        return EnumDeserializer(
            enm=self.__enm,
            aliases=_aliases,
        )

    def coerce(
        self,
        raw: RawWrapper,
        diagnostics: Diagnostics,
        from_lut: CheckLutFn[T],
    ) -> Result[T]:
        parsed = raw.as_str()
        if parsed is None:
            diagnostics.push_error(DeserializerError.create_error("Expected enum"))
            return Result.failed()
        if parsed in self.__value_aliases:
            parsed = self.__value_aliases[parsed]

        try:
            parsed_item = self.__enm(parsed)
            return Result.from_value(parsed_item)
        except Exception as e:
            diagnostics.push_error(
                DeserializerError.create_error(f"Invalid enum {parsed}")
            )
            return Result.failed()

File: clients/ts/src/baml_lib/deserializer/enum_deserializer.ts
import { BaseDeserializer, CheckLutFn, Result } from "./base_deserializer";
import { Diagnostics } from "./diagnostics";
import { RawWrapper } from "./raw_wrapper/raw_wrapper";
import { JSONSchema7 } from "json-schema";

class EnumDeserializer<T extends Record<string, string>> extends BaseDeserializer<keyof T> {
    private aliases: Map<string, string> = new Map();

    private constructor(public readonly name: string, private readonly values: Map<string, keyof T>, aliases: Record<string, string>) {
        super(3);
        
        // Aliases are case-insensitive
        Object.entries(aliases).forEach(([k, v]) => {
            this.aliases.set(k.toLowerCase(), v.toLowerCase());
        });
    }

    static from_schema<T extends Record<string, string>>(schema: JSONSchema7, aliases: Record<string, string>): EnumDeserializer<T> {
        if (schema.enum === undefined) {
            throw new Error(`Schema must be an enum`);
        }

        const name = schema.title ?? "unnamed_enum";
        const values = schema.enum.map(x => {
            if (x === null) {
                throw new Error(`Null schema not supported for value ${x}`);
            }
            if (typeof x === "boolean") {
                throw new Error(`Boolean schema not supported for value ${x}`);
            }
            if (typeof x === "number") {
                throw new Error(`Number schema not supported for value ${x}`);
            }
            if (Array.isArray(x)) {
                throw new Error(`Array schema not supported for value ${x}`);
            }

            if (typeof x === "string") {
                return x as keyof T & string;
            }

            const value = x.const;
            if (value === undefined || typeof value !== "string") {
                throw new Error(`Invalid schema for value ${x}`);
            }
            return value as keyof T & string;
        });

        return new EnumDeserializer(name, new Map(values.map(v => [v.toLowerCase(), v])), aliases);
    }

    copy_with_aliases(aliases: Record<string, string>): EnumDeserializer<T> {
        this.aliases.forEach((v, k) => {
            if (!aliases[k]) {
                aliases[k] = v;
            }
        });
        return new EnumDeserializer(this.name, this.values, aliases);
    }

    coerce(raw: RawWrapper, diagnostics: Diagnostics, fromLut: CheckLutFn<any>): Result<keyof T> {
        diagnostics.pushScope(this.name);
        const parsed = raw.as_smart_str(false);
        if (parsed === undefined) {
            diagnostics.pushUnknownError(`Expected string, got ${raw.as_self()}`);
            diagnostics.popScope(false);
            return Result.failed();
        }

        const value = parsed.toLowerCase();
        const valName = this.values.get(value);
        if (!valName) {
            diagnostics.pushUnknownError(`Unknown value: ${parsed}`);
            diagnostics.popScope(false);
            return Result.failed();
        }
        diagnostics.popScope(true);

        return Result.from_value(valName);
    }
}

export { EnumDeserializer };